### PR TITLE
Use real laziness for Re_automata.state

### PR DESCRIPTION
**Pull request by rgrinberg**
Also, since it's just an accessor, it should rather be
Re_automata.State.t

The benchmarks results are the same except for 1 benchmark which is about %5
slower.

### CODE REVIEW

**Comment by rgrinberg** [comment]
```
.----------------------------------+--------------+-------------------+--------.
| Name                             | Time(master) | Time(status-lazy) | Delta  |
+----------------------------------+--------------+-------------------+--------+
| 20 zeroes/exec/case 0            | 173.16ns     | 165.99ns          | -4.14% |
| 20 zeroes/execp/case 0           | 132.28ns     | 134.54ns          | 1.71%  |
| 20 zeroes/exec_opt/case 0        | 172.33ns     | 180.17ns          | 4.55%  |
| lots of a's/exec/case 0          | 438.09ns     | 432.49ns          | -1.28% |
| lots of a's/execp/case 0         | 454.01ns     | 459.55ns          | 1.22%  |
| lots of a's/exec_opt/case 0      | 433.77ns     | 451.06ns          | 3.99%  |
| media type match/exec/case 0     | 84.80ns      | 85.12ns           | 0.38%  |
| media type match/execp/case 0    | 61.61ns      | 62.08ns           | 0.76%  |
| media type match/exec_opt/case 0 | 85.01ns      | 85.08ns           | 0.08%  |
| uri/exec/case 0                  | 149.38ns     | 146.06ns          | -2.22% |
| uri/exec/case 1                  | 237.89ns     | 243.33ns          | 2.29%  |
| uri/exec/case 2                  | 143.67ns     | 142.45ns          | -0.85% |
| uri/execp/case 0                 | 124.04ns     | 125.05ns          | 0.81%  |
| uri/execp/case 1                 | 227.65ns     | 227.75ns          | 0.04%  |
| uri/execp/case 2                 | 114.72ns     | 120.91ns          | 5.40%  |
| uri/exec_opt/case 0              | 147.10ns     | 145.29ns          | -1.23% |
| uri/exec_opt/case 1              | 236.22ns     | 236.16ns          | -0.03% |
| uri/exec_opt/case 2              | 143.55ns     | 142.10ns          | -1.01% |
'----------------------------------+--------------+-------------------+--------'
```

File: lib/re_automata.ml [hunk 1]
```diff
@@ -307,25 +307,41 @@ module State = struct
...
-    ; status = None
-    ; hash = hash idx cat desc}
+  let mk idx category desc =
+    let rec t =
+      { idx
+      ; category
+      ; desc
+      ; status
+      ; hash = hash idx category desc}
+    and status = lazy (mk_status t.desc) in
```
Comment (Drup) [inline lib/re_automata.ml]: Why the rec, instead of  `{idx ; category ; desc ; status = lazy (mk_status desc) ; hash = hash idx category desc }` ?

Comment (rgrinberg) [inline lib/re_automata.ml]: I thought that in my version if the desc is updated before status is forced, we'll be using the updated version of desc to make status. This is too much the semantics of the existing status more closely. I could be wrong though.

Comment (Drup) [inline lib/re_automata.ml]: But the `desc` field itself is not mutable, so that doesn't change anything.

**Comment by rgrinberg** [comment]
@Drup you're right, I must have been hallucinating.

**Pull request closed without merging**